File: backend/booking-service/src/models/booking.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import datetime
from enum import Enum

class BookingStatus(str, Enum):
    PENDING = "pending"
    CONFIRMED = "confirmed"
    SEATED = "seated"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    NO_SHOW = "no_show"

class BookingType(str, Enum):
    REGULAR = "regular"
    WALK_IN = "walk_in"
    VIP = "vip"
    GROUP = "group"

class TableStatus(str, Enum):
    AVAILABLE = "available"
    RESERVED = "reserved"
    OCCUPIED = "occupied"
    CLEANING = "cleaning"
    MAINTENANCE = "maintenance"

class Table(BaseModel):
    id: str
    restaurant_id: str
    table_number: str
    capacity: int = Field(..., gt=0)
    status: TableStatus = TableStatus.AVAILABLE
    location: Optional[str] = None
    features: List[str] = []
    created_at: datetime
    updated_at: datetime

class BookingBase(BaseModel):
    restaurant_id: str
    customer_id: str
    table_id: str
    booking_type: BookingType = BookingType.REGULAR
    party_size: int = Field(..., gt=0)
    booking_date: datetime
    start_time: datetime
    end_time: datetime
    special_requests: Optional[str] = None
    contact_phone: str
    contact_email: Optional[str] = None
    status: BookingStatus = BookingStatus.PENDING

    @validator('end_time')
    def end_time_must_be_after_start_time(cls, v, values):
        if 'start_time' in values and v <= values['start_time']:
            raise ValueError('end_time must be after start_time')
        return v

    @validator('party_size')
    def party_size_must_be_valid(cls, v, values):
        if 'table_id' in values:
            # Here you would typically check against the table's capacity
            # For now, we'll just ensure it's positive
            if v <= 0:
                raise ValueError('party_size must be positive')
        return v

class Booking(BookingBase):
    id: str
    booking_number: str
    created_at: datetime
    updated_at: datetime
    confirmed_at: Optional[datetime] = None
    seated_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None
    no_show_at: Optional[datetime] = None
    cancellation_reason: Optional[str] = None
    waitlist_position: Optional[int] = None
    waitlist_joined_at: Optional[datetime] = None
    waitlist_notified_at: Optional[datetime] = None

class BookingCreate(BookingBase):
    pass

class BookingUpdate(BaseModel):
    status: Optional[BookingStatus] = None
    party_size: Optional[int] = Field(None, gt=0)
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    special_requests: Optional[str] = None
    contact_phone: Optional[str] = None
    contact_email: Optional[str] = None
    cancellation_reason: Optional[str] = None

class BookingFilter(BaseModel):
    restaurant_id: Optional[str] = None
    customer_id: Optional[str] = None
    status: Optional[BookingStatus] = None
    booking_type: Optional[BookingType] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    table_id: Optional[str] = None

class BookingStats(BaseModel):
    total_bookings: int
    confirmed_bookings: int
    cancelled_bookings: int
    no_shows: int
    average_party_size: float
    bookings_by_status: dict
    bookings_by_type: dict
    peak_hours: List[dict]
    popular_tables: List[dict]
    waitlist_stats: Optional[dict] = None 

File: devops/ci-cd/lib copy/models/booking.dart

enum BookingStatus {
  pending,
  confirmed,
  cancelled,
  completed,
  noShow
}

enum BookingType {
  regular,
  special,
  group
}

class Booking {
  final String id;
  final String restaurantId;
  final String customerId;
  final String tableId;
  final BookingType type;
  final int partySize;
  final DateTime bookingDate;
  final DateTime startTime;
  final DateTime endTime;
  final String? specialRequests;
  final String contactPhone;
  final String? contactEmail;
  final BookingStatus status;
  final DateTime createdAt;
  final DateTime updatedAt;

  Booking({
    required this.id,
    required this.restaurantId,
    required this.customerId,
    required this.tableId,
    required this.type,
    required this.partySize,
    required this.bookingDate,
    required this.startTime,
    required this.endTime,
    this.specialRequests,
    required this.contactPhone,
    this.contactEmail,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'],
      restaurantId: json['restaurant_id'],
      customerId: json['customer_id'],
      tableId: json['table_id'],
      type: BookingType.values.firstWhere(
        (e) => e.toString().split('.').last == json['booking_type'],
      ),
      partySize: json['party_size'],
      bookingDate: DateTime.parse(json['booking_date']),
      startTime: DateTime.parse(json['start_time']),
      endTime: DateTime.parse(json['end_time']),
      specialRequests: json['special_requests'],
      contactPhone: json['contact_phone'],
      contactEmail: json['contact_email'],
      status: BookingStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['status'],
      ),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'restaurant_id': restaurantId,
      'customer_id': customerId,
      'table_id': tableId,
      'booking_type': type.toString().split('.').last,
      'party_size': partySize,
      'booking_date': bookingDate.toIso8601String(),
      'start_time': startTime.toIso8601String(),
      'end_time': endTime.toIso8601String(),
      'special_requests': specialRequests,
      'contact_phone': contactPhone,
      'contact_email': contactEmail,
      'status': status.toString().split('.').last,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

class Table {
  final String id;
  final String restaurantId;
  final String tableNumber;
  final int capacity;
  final String status;
  final String? location;
  final List<String> features;
  final DateTime createdAt;
  final DateTime updatedAt;

  Table({
    required this.id,
    required this.restaurantId,
    required this.tableNumber,
    required this.capacity,
    required this.status,
    this.location,
    required this.features,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Table.fromJson(Map<String, dynamic> json) {
    return Table(
      id: json['id'],
      restaurantId: json['restaurant_id'],
      tableNumber: json['table_number'],
      capacity: json['capacity'],
      status: json['status'],
      location: json['location'],
      features: List<String>.from(json['features']),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'restaurant_id': restaurantId,
      'table_number': tableNumber,
      'capacity': capacity,
      'status': status,
      'location': location,
      'features': features,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
} 